odilo.py

```diff
@@ -0,0 +1,598 @@
+
+    def __init__(self, _db, collection):
+        if collection.protocol != ExternalIntegration.ODILO:
+            raise ValueError("Collection protocol is %s, but passed into OdiloAPI!" % collection.protocol)
+
+        self._db = _db
+        self.analytics = Analytics(self._db)
+
+        self.collection_id = collection.id
+        self.token = None
+        self.client_key = collection.external_integration.username
+        self.client_secret = collection.external_integration.password
+        self.library_api_base_url = collection.external_integration.setting(self.LIBRARY_API_BASE_URL).value
+
+        if not self.client_key or not self.client_secret or not self.library_api_base_url:
+            raise CannotLoadConfiguration("Odilo configuration is incomplete.")
+
+        # Use utf8 instead of unicode encoding
+        settings = [self.client_key, self.client_secret, self.library_api_base_url]
+        self.client_key, self.client_secret, self.library_api_base_url = (
+            setting.encode('utf8') for setting in settings
+        )
+
+        # Get set up with up-to-date credentials from the API.
+        self.check_creds()
+        if not self.token:
+            raise CannotLoadConfiguration("Invalid credentials for %s, cannot intialize API %s"
+                                          % (self.client_key, self.library_api_base_url))
+
+    @property
+    def collection(self):
+        return Collection.by_id(self._db, id=self.collection_id)
+
+    @property
+    def source(self):
+        return DataSource.lookup(self._db, DataSource.ODILO)
+
+    def check_creds(self, force_refresh=False):
+        """If the Bearer Token has expired, update it."""
+        if force_refresh:
+            refresh_on_lookup = lambda x: x
+        else:
+            refresh_on_lookup = self.refresh_creds
+
+        credential = self.credential_object(refresh_on_lookup)
+        if force_refresh:
+            self.refresh_creds(credential)
+        self.token = credential.credential
+
+    def credential_object(self, refresh):
+        """Look up the Credential object that allows us to use
+        the Odilo API.
+        """
+        return Credential.lookup(self._db, DataSource.ODILO, None, None, refresh)
+
+    def refresh_creds(self, credential):
+        """Fetch a new Bearer Token and update the given Credential object."""
+        try:
+            response = self.token_post(
+                self.TOKEN_ENDPOINT,
+                dict(grant_type="client_credentials"),
+                allowed_response_codes=[200]
+            )
+            data = response.json()
+        except (HTTPException, RequestNetworkException) as e:
+            self.log.error("Cannot connect with %s, error: %s" % (self.TOKEN_ENDPOINT, e.message))
+            return None
+
+        if response.status_code == 200:
+            self._update_credential(credential, data)
+            self.token = credential.credential
+            return 'OK'
```

Comment:
I see that you're checking this value in the calling method in `circulation`. I don't like the idea of passing specific string values back and forth. I would rather do one of the following:

* Return None on success and an error string if there's an error (this would mean changing line 155 above).
* Raise an exception if there's an error; return value is irrelevant.

Reply:
Ok, I have removed the 'return' clauses and raised an exception if needed instead.